packages/runtime/runtime-definitions/src/dataStoreRegistry.ts

```diff
@@ -21,12 +21,23 @@ export type FluidDataStoreRegistryEntry = Readonly<
  * @alpha
  */
 export type NamedFluidDataStoreRegistryEntry = [string, Promise<FluidDataStoreRegistryEntry>];
+
+/**
+ * An associated pair of an identifier and registry entry.  Registry entries
+ * may be dynamically loaded.
+ * @legacy
+ * @alpha
+ */
+export type NamedFluidDataStoreRegistryEntry2 = [
```

Comment:
adding a new type is the most compatible thing according to our type tests, as it typescript has trouble with inference over typed arrays. not sure what to call this, everything else feels quite verbose, so kind of happy with just adding 2

Reply:
The name does matter to the registry owner. Specifically, the ContainerRuntime needs to know the names of the factories it wants to create, and DataStore's need to know the names of the of the factories they want to create. This is further complicated by the fact that most datastore creations are driven from the ContainerRuntime  with an explicit path of names, so they need to be fixed. I've made it a bit better in the PR, as createChildDataStoreSync leverages the path of the current context internally, and just appends the sub-factories name, so it doesn't need to care about the name assigned in the ContainerRuntime's registry.

Reply:
I'm talking about the factory being opinionated about its own name, not the registry owner being opinionated about the names of the factories in its registry.  The latter is of course important as you point out, but the former is a net negative.

Reply:
I think the whole pattern should go away (not in this PR, of course):
1. It's important that we easily support delay-loading. The only way to get there today is by using LazyPromises - I think that's wrong. Instead, everywhere where we assume delay-loading, we should use async methods / functions, not promises.
2. There is really no need for extra interface / contract to be part of our API. For example, ContainerRuntime should not take NamedFluidDataStoreRegistryEntries, it should take IFluidDataStoreRegistry, and let the caller (if they want to) use something like NamedFluidDataStoreRegistryEntries and wrap it via FluidDataStoreRegistry. Same ideally with data store registries.
3. The problem that Matt raises - it's not clear (from existing interfaces) what is the contract today RE who owns the type name - factory or the registry (i.e. can I, as implementer of the registry, give my own names? If yes, why factory has a type?).